### PR TITLE
GH-27 Applications List

### PR DESCRIPTION
**Pull request by mskwara**
- Created Applications List
 - Refactored all styles.css.ts to export named export (we can use easily `styles` variable with no need to manually change the imported variable name)
 - Created `generate` page for DEV environment. This way the database is not regenerated every time the page is refreshed, but only when we intentionally go to `/generate` URL
 - Added mapper to useDatabase hook. This way e.g. dates will be transformed to Date object in the hook and will be ready to use in components.

### CODE REVIEW

File: src/pages/Applications/List/index.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import ApplicationTile from '~/components/ApplicationTile';
+import Loader from '~/components/Loader';
+import { useApplicationsQuery } from './queries';
+import { styles } from './styles.css';
+
+const ApplicationsList = () => {
```
Comment (jakobos) [inline src/pages/Applications/List/index.tsx]: Maybe you could extract query logic from application list? Application list should be stupid and it shouldn't know anything about query logic.

So what do you think about:
1. Add `ListLoader` component for loading data and handling statuses like `isLoading` or case when there is no data
```
type ListLoaderResult = {
  applications: Application[];
};

type Props = {
  children: (result: ListLoaderResult) => JSX.Element;
};

const NotFoundMessage = () => <div>No data found</div>;

export const ListLoader = (props: Props): JSX.Element => {
  const { data, isLoading } = useApplicationsQuery();

  if (isLoading) {
    return <Loader />;
  }
  if (!data) {
    return <NotFoundMessage />;
  }
  return props.children({ applications: data });
};
```
2. Next, there could be `ApplicationList` component which gets data via props and knows nothing about query logic
```
type Props = {
  applications: Application[];
};

export const ApplicationsList = ({ applications }: Props) => {
  return (
    <div className={styles.root}>
      <div className={styles.options}>TODO: Options</div>
      <main className={styles.list}>
        {applications.map((application) => (
          <ApplicationTile
            key={application.id}
            companyName={application.companyName}
            roleName={application.roleName}
            status={application.status}
            salary={{
              from: application.salaryRange.from,
              to: application.salaryRange.to,
            }}
            linkToOffer={application.linkToOffer}
            createdAt={application.createdAt}
          />
        ))}
      </main>
    </div>
  );
};
```

3. In the end there could be something like `ApplicationsListSection` - the place of usage loader and list:
```
const ApplicationsListSection = () => {
  return (
    <ListLoader>
      {({ applications }) => <ApplicationsList applications={applications} />}
    </ListLoader>
  );
};
```
I think it looks like more elegant and less complex :)

File: src/hooks/useDatabase/utils.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { Application, Collection, ModelBase } from '~/@types/common';
+
+export const mapEntry = <T extends ModelBase>(collection: Collection, entry: T): T => {
+  const mappedBase = {
+    ...entry,
+    createdAt: new Date(entry.createdAt),
```
Comment (jakobos) [inline src/hooks/useDatabase/utils.ts:10]: `entry.createdAt` is type of Date so why you wrap it again into another Date ?

File: src/pages/Applications/List/index.tsx (hunk 1 above)
Comment (mskwara) [reply src/pages/Applications/List/index.tsx]: Please check conversation on Slack.
I also pushed changes which show my idea (check Slack!)
I'm not sure about filenames, but ... check Slack!

File: src/hooks/useDatabase/utils.ts (hunk 2 above)
Comment (mskwara) [reply src/hooks/useDatabase/utils.ts:10]: You are right! I simplified the fact that Local Storage stores dates as strings.
I pushed changes that add `DBModelBase` and `DBApplication` types. I had to change types a little bit in `useDatabase`.
Also fixed bug in `updateById` function - I mapped the entry twice (first time in `get` function and next in `updateById`) - thanks for spotting the issue with types!

File: src/hooks/useDatabase/index.ts [hunk 3]
```diff
@@ -1,34 +1,41 @@
...
+import { DBApplication, DBModelBase } from '~/@types/database';
+import { mapEntry } from './utils';
 
 type DropFirst<T extends unknown[]> = T extends [any, ...infer U] ? U : never;
 
 const FREE_ID_KEY = 'free-id';
 
 const useDatabase = () => {
-  const get = <T extends ModelBase>(
+  const get = <Model extends ModelBase, DBModel extends DBModelBase>(
```
Comment (jakobos) [inline src/hooks/useDatabase/index.ts]: Please use uppercase names as name of generic type, like: `MODEL` or `DB_MODEL` - it is easiest to detect that this is a generic type

Comment (mskwara) [reply src/hooks/useDatabase/index.ts]: Done

**Review by jakobos — APPROVED** [review]

**Pull request merged**